src/main/java/org/folio/innreach/domain/service/impl/InnReachTransactionActionServiceImpl.java

```diff
@@ -129,30 +134,46 @@ public void associateNewLoanWithTransaction(LoanDTO loan) {
 
   @Override
   public void handleLoanUpdate(LoanDTO loan) {
+    InnReachTransaction associatedTransactionWithLoan = null;
+    Optional<InnReachTransaction> optionalTransaction = transactionRepository.fetchOneByLoanId(loan.getId());
+    if (optionalTransaction.isPresent()) {
+      associatedTransactionWithLoan = optionalTransaction.get();
+    }
+
     if ("checkedin".equals(loan.getAction()) && "Closed".equalsIgnoreCase(loan.getStatus().getName())) {
-      updateAssociatedTransaction(loan, transaction -> {
-        log.info("Updating transaction {} on loan closure {}", transaction.getId(), loan.getId());
-        transaction.getHold().setDueDateTime(null);
-        transaction.setState(ITEM_IN_TRANSIT);
-        reportItemInTransit(transaction);
-      });
+      if (associatedTransactionWithLoan != null && (associatedTransactionWithLoan.getType() == ITEM)) {
+        log.info("Updating transaction {} on loan to final check-in {}", associatedTransactionWithLoan.getId(), loan.getId());
+        associatedTransactionWithLoan.getHold().setDueDateTime(null);
+        var transactionHold = (TransactionHold) associatedTransactionWithLoan.getHold();
+        var transactionItemHold = (TransactionItemHold) associatedTransactionWithLoan.getHold();
+        transactionItemHold.setPatronName(null);
+        transactionHold.setPatronId(null);
+        associatedTransactionWithLoan.setState(FINAL_CHECKIN);
+        reportFinalCheckIn(associatedTransactionWithLoan);
+        transactionRepository.save(associatedTransactionWithLoan);
+      }
+      else {
+        updateAssociatedTransaction(loan, transaction -> {
+          log.info("Updating transaction {} on loan closure {}", transaction.getId(), loan.getId());
+          transaction.getHold().setDueDateTime(null);
+          transaction.setState(ITEM_IN_TRANSIT);
+          reportItemInTransit(transaction);
+        });
```

Comment:
remove this call `updateAssociatedTransaction` and reuse `associatedTransactionWithLoan`

Reply:
Done

Reply:
Fixed